### DISCUSSION TITLE
Is SetDllImportResolver invoked multiple times normal?

### DISCUSSION THREAD

**Discussion by nathan130200** [dotnet/runtime#86088] [Q&A] [answered]
I'm working on bindings for Expat to implement my own XMPP library. Then i provide a way to set an environment variable to manually point to correct expat library. Everything is working as excepted, expat is begin called and parsing is working, but...

I had placed an `Debug.WriteLine` to see what happens here:
```cs
    static nint TryResolveExpat(string name, Assembly assembly, DllImportSearchPath? searchPath = default)
    {
        if (name.Contains("libexpat", StringComparison.InvariantCultureIgnoreCase))
        {
            string libFile = Environment.GetEnvironmentVariable("EXPAT_LIBRARY_DIR");

            // fallback to default dll importer
            if (string.IsNullOrWhiteSpace(libFile))
                return default;

            // fallback to default dll importer
            if (!File.Exists(libFile))
                return default;

            Debug.WriteLine("Expat DLL Import Resolved: " + libFile);

            // attempt to load expat provided in env.
            return NativeLibrary.Load(libFile, assembly, searchPath);
        }

        // fallback to default dll importer
        return nint.Zero;
    }
```

And for my surprise is calling multiple times, my guess this is causing some memory leak loading many times into memory?

```cs
Expat DLL Import Resolved: C:\vcpkg\installed\x64-windows\debug\bin\libexpatd.dll
Using libexpat v2.5.0!
Expat DLL Import Resolved: C:\vcpkg\installed\x64-windows\debug\bin\libexpatd.dll
Expat DLL Import Resolved: C:\vcpkg\installed\x64-windows\debug\bin\libexpatd.dll
Expat DLL Import Resolved: C:\vcpkg\installed\x64-windows\debug\bin\libexpatd.dll
Expat DLL Import Resolved: C:\vcpkg\installed\x64-windows\debug\bin\libexpatd.dll
Expat DLL Import Resolved: C:\vcpkg\installed\x64-windows\debug\bin\libexpatd.dll
Expat DLL Import Resolved: C:\vcpkg\installed\x64-windows\debug\bin\libexpatd.dll
Expat DLL Import Resolved: C:\vcpkg\installed\x64-windows\debug\bin\libexpatd.dll
Expat DLL Import Resolved: C:\vcpkg\installed\x64-windows\debug\bin\libexpatd.dll
Expat DLL Import Resolved: C:\vcpkg\installed\x64-windows\debug\bin\libexpatd.dll
Expat DLL Import Resolved: C:\vcpkg\installed\x64-windows\debug\bin\libexpatd.dll
```

The file in question 
[PInvoke.cs](https://github.com/nathan130200/XmppSharp/blob/96e62de7518d4f5ad39d1fb463bb6f8d0736182d/XmppSharp/Expat/PInvoke.cs#L32-L54)

1st: Get expat version (then next line you can see the version)
2nd: Is to invoke `XML_CreateParser` and return the parser pointer from unmanaged.
3rd~7th: calls is to set start/end handler, cdata open/close handler, comment handler and text handler. 
8th... calls is sending data to parser (small chunks of 256 bytes for testing).

<hr/>

**UPDATE**: I've updated code to display address of returned `nint` pointer and its point to same library handle.

```cs
Expat DLL Import Resolved: C:\vcpkg\installed\x64-windows\debug\bin\libexpatd.dll (at: 0000007FFC41FE0000)
Using libexpat v2.5.0!
Expat DLL Import Resolved: C:\vcpkg\installed\x64-windows\debug\bin\libexpatd.dll (at: 0000007FFC41FE0000)
Expat DLL Import Resolved: C:\vcpkg\installed\x64-windows\debug\bin\libexpatd.dll (at: 0000007FFC41FE0000)
Expat DLL Import Resolved: C:\vcpkg\installed\x64-windows\debug\bin\libexpatd.dll (at: 0000007FFC41FE0000)
Expat DLL Import Resolved: C:\vcpkg\installed\x64-windows\debug\bin\libexpatd.dll (at: 0000007FFC41FE0000)
Expat DLL Import Resolved: C:\vcpkg\installed\x64-windows\debug\bin\libexpatd.dll (at: 0000007FFC41FE0000)
Expat DLL Import Resolved: C:\vcpkg\installed\x64-windows\debug\bin\libexpatd.dll (at: 0000007FFC41FE0000)
Expat DLL Import Resolved: C:\vcpkg\installed\x64-windows\debug\bin\libexpatd.dll (at: 0000007FFC41FE0000)
Expat DLL Import Resolved: C:\vcpkg\installed\x64-windows\debug\bin\libexpatd.dll (at: 0000007FFC41FE0000)
Expat DLL Import Resolved: C:\vcpkg\installed\x64-windows\debug\bin\libexpatd.dll (at: 0000007FFC41FE0000)
Expat DLL Import Resolved: C:\vcpkg\installed\x64-windows\debug\bin\libexpatd.dll (at: 0000007FFC41FE0000)
```

But still this normal calling multipletimes even the library was resolved once?

**Comment by gfoidl** [comment]
@jkotas please have a look here.

**Comment by jkotas** [comment, marked as answer]
Yes, this is expected. Runtime does not maintain caches of resolved native dll handles.

**Reply by gfoidl** [reply]
Thanks.

For the record from https://github.com/dotnet/runtime/issues/46463#issuecomment-752350722
> It is ok to cache this pointer, as long as your code does not call `NativeLibrary.Free` on it. The runtime itself does not do any lifetime management with `DllImportResolver`, it assumes that your code will keep the library loaded long enough.

---

Should this non-caching behavior put into the docs too?
At the moment it's not clear from the docs whether the runtime caches the result or not w/o digging through issues here.

**Reply by jkotas** [reply]
Remarks section in https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.nativelibrary.setdllimportresolver says that the runtime does not do any lifetime management. It suggests that it won't cache the handles. If you would like to see caching to be explicitly mentioned for clarity, feel free to submit a PR to http://github.com/dotnet/dotnet-api-docs/ .

**Reply by gfoidl** [reply]
Got it on https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.dllimportresolver. Thanks!

-> https://github.com/dotnet/dotnet-api-docs/pull/11864
